[Remote State] NullPointerException when remote publication enabled with remote state disabled

### Describe the bug

If remote state is disabled when node has `cluster.remote_store.state.enabled: false` in opensearch.yml but remote publication is also enabled with following setting `opensearch.experimental.feature.remote_store.publication.enabled=true`. The cluster manager tries to publish the state but encounter null pointer exception as remote state is not enabled.

### Related component

Cluster Manager

### To Reproduce

1. Add following settings to the `opensearch.yml` file
```
node.attr.remote_store.segment.repository: my-fs-repository
node.attr.remote_store.translog.repository: my-fs-repository
node.attr.remote_store.routing_table.repository: my-fs-repository
node.attr.remote_store.repository.my-fs-repository.type: fs
node.attr.remote_store.repository.my-fs-repository.settings.location: ~/os_data/repos/repo-1
cluster.remote_store.state.enabled: true
node.attr.remote_store.state.repository: my-fs-repository
```
2. Run the opensearch process with following settings
```
OPENSEARCH_JAVA_OPTS="-Dopensearch.experimental.feature.remote_store.enabled=true -Dopensearch.experimental.feature.replication_type.enabled=true -Dopensearch.experimental.feature.remote_store.routing.enabled=true -Dopensearch.experimental.feature.remote_store.publication.enabled=true -Daws.region=us-east-1" ./build/distribution/local/opensearch-3.0.0-SNAPSHOT/bin/opensearch -E cluster.name=hishiv-cluster -E path.data=~/os_data/master1 -E path.repo=~/os_data/repos -E node.name=master1 -E node.master=true -E node.data=false -E node.ingest=false  -E cluster.initial_master_nodes=master1
```
3. Publication failing with following error
```
[2024-08-09T15:44:24,322][WARN ][o.o.c.c.PublicationTransportHandler] [master1] error sending remote cluster state to {master1}{ldGqZarCS3K9MBuR7idMlQ}{U9B1LEitTMKYs2hpP8S2vA}{127.0.0.1}{127.0.0.1:9300}{mr}{shard_indexing_pressure_enabled=true}
java.lang.NullPointerException: Cannot invoke "org.opensearch.gateway.GatewayMetaState$RemotePersistedState.getLastUploadedManifestFile()" because the return value of "org.opensearch.cluster.coordination.PersistedStateRegistry.getPersistedState(org.opensearch.cluster.coordination.PersistedStateRegistry$PersistedStateType)" is null
	at org.opensearch.cluster.coordination.PublicationTransportHandler$PublicationContext.sendRemoteClusterState(PublicationTransportHandler.java:527) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.PublicationTransportHandler$PublicationContext.sendPublishRequest(PublicationTransportHandler.java:474) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.Coordinator$CoordinatorPublication.sendPublishRequest(Coordinator.java:1843) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.Publication$PublicationTarget.sendPublishRequest(Publication.java:287) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596) [?:?]
	at org.opensearch.cluster.coordination.Publication.start(Publication.java:94) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.Coordinator.publish(Coordinator.java:1356) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.MasterService.publish(MasterService.java:385) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.MasterService.runTasks(MasterService.java:367) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.MasterService$Batcher.run(MasterService.java:229) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:210) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:252) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:923) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedOpenSearchThreadPoolExecutor.java:283) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedOpenSearchThreadPoolExecutor.java:246) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
[2024-08-09T15:44:24,326][WARN ][o.o.c.s.MasterService    ] [master1] failing [Tasks batched with key: org.opensearch.cluster.coordination.JoinHelper and count: 3]: failed to commit cluster state version [1]
org.opensearch.cluster.coordination.FailedToCommitClusterStateException: publishing failed
	at org.opensearch.cluster.coordination.Coordinator.publish(Coordinator.java:1360) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.MasterService.publish(MasterService.java:385) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.MasterService.runTasks(MasterService.java:367) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.MasterService$Batcher.run(MasterService.java:229) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:210) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:252) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:923) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedOpenSearchThreadPoolExecutor.java:283) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedOpenSearchThreadPoolExecutor.java:246) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.lang.ClassCastException: class java.lang.NullPointerException cannot be cast to class org.opensearch.transport.TransportException (java.lang.NullPointerException is in module java.base of loader 'bootstrap'; org.opensearch.transport.TransportException is in unnamed module of loader 'app')
	at org.opensearch.cluster.coordination.Publication$PublicationTarget$PublishResponseHandler.onFailure(Publication.java:410) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.Coordinator$5.onFailure(Coordinator.java:1403) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.PublicationTransportHandler$PublicationContext$1.onFailure(PublicationTransportHandler.java:465) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.PublicationTransportHandler$PublicationContext.sendRemoteClusterState(PublicationTransportHandler.java:571) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.PublicationTransportHandler$PublicationContext.sendPublishRequest(PublicationTransportHandler.java:474) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.Coordinator$CoordinatorPublication.sendPublishRequest(Coordinator.java:1843) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.Publication$PublicationTarget.sendPublishRequest(Publication.java:287) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596) ~[?:?]
	at org.opensearch.cluster.coordination.Publication.start(Publication.java:94) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.opensearch.cluster.coordination.Coordinator.publish(Coordinator.java:1356) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	... 11 more
```

### Expected behavior

If we are in such inconsistent state where the remote publication is enabled without remote state being enabled. We should fall back to publication over transport call. 

### Additional Details

_No response_

--- a/server/src/internalClusterTest/java/org/opensearch/gateway/remote/RemoteStatePublicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/gateway/remote/RemoteStatePublicationIT.java
@@ -50,15 +50,22 @@ import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
 public class RemoteStatePublicationIT extends RemoteStoreBaseIntegTestCase {
 
     private static String INDEX_NAME = "test-index";
+    private boolean isRemoteStateEnabled = true;
+    private String isRemotePublicationEnabled = "true";
 
     @Before
     public void setup() {
         asyncUploadMockFsRepo = false;
+        isRemoteStateEnabled = true;
+        isRemotePublicationEnabled = "true";
     }
 
     @Override
     protected Settings featureFlagSettings() {
-        return Settings.builder().put(super.featureFlagSettings()).put(FeatureFlags.REMOTE_PUBLICATION_EXPERIMENTAL, "true").build();
+        return Settings.builder()
+            .put(super.featureFlagSettings())
+            .put(FeatureFlags.REMOTE_PUBLICATION_EXPERIMENTAL, isRemotePublicationEnabled)
+            .build();
     }
 
     @Override
@@ -76,7 +83,7 @@ public class RemoteStatePublicationIT extends RemoteStoreBaseIntegTestCase {
         );
         return Settings.builder()
             .put(super.nodeSettings(nodeOrdinal))
-            .put(REMOTE_CLUSTER_STATE_ENABLED_SETTING.getKey(), true)
+            .put(REMOTE_CLUSTER_STATE_ENABLED_SETTING.getKey(), isRemoteStateEnabled)
             .put("node.attr." + REMOTE_STORE_ROUTING_TABLE_REPOSITORY_NAME_ATTRIBUTE_KEY, routingTableRepoName)
             .put(routingTableRepoTypeAttributeKey, ReloadableFsRepository.TYPE)
             .put(routingTableRepoSettingsAttributeKeyPrefix + "location", segmentRepoPath)
@@ -136,6 +143,18 @@ public class RemoteStatePublicationIT extends RemoteStoreBaseIntegTestCase {
         }
     }
 
+    public void testRemotePublicationDisableIfRemoteStateDisabled() {
+        // only disable remote state
+        isRemoteStateEnabled = false;
+        // create cluster with multi node with in-consistent settings
+        prepareCluster(3, 2, INDEX_NAME, 1, 2);
+        // assert cluster is stable, ensuring publication falls back to legacy transport with inconsistent settings
+        ensureStableCluster(5);
+        ensureGreen(INDEX_NAME);
+
+        assertNull(internalCluster().getCurrentClusterManagerNodeInstance(RemoteClusterStateService.class));
+    }
+
     private Map<String, Integer> getMetadataFiles(BlobStoreRepository repository, String subDirectory) throws IOException {
         BlobPath metadataPath = repository.basePath()
             .add(
@@ -151,5 +170,4 @@ public class RemoteStatePublicationIT extends RemoteStoreBaseIntegTestCase {
             return fileName.split(DELIMITER)[0];
         }).collect(Collectors.toMap(Function.identity(), key -> 1, Integer::sum));
     }
-
 }

--- a/server/src/main/java/org/opensearch/cluster/coordination/CoordinationState.java
+++ b/server/src/main/java/org/opensearch/cluster/coordination/CoordinationState.java
@@ -105,7 +105,8 @@ public class CoordinationState {
             .getLastAcceptedConfiguration();
         this.publishVotes = new VoteCollection();
         this.isRemoteStateEnabled = isRemoteStoreClusterStateEnabled(settings);
-        this.isRemotePublicationEnabled = FeatureFlags.isEnabled(REMOTE_PUBLICATION_EXPERIMENTAL)
+        this.isRemotePublicationEnabled = isRemoteStateEnabled
+            && FeatureFlags.isEnabled(REMOTE_PUBLICATION_EXPERIMENTAL)
             && localNode.isRemoteStatePublicationEnabled();
     }
 

--- a/server/src/test/java/org/opensearch/cluster/coordination/CoordinationStateTests.java
+++ b/server/src/test/java/org/opensearch/cluster/coordination/CoordinationStateTests.java
@@ -66,6 +66,8 @@ import org.mockito.Mockito;
 
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.emptySet;
+import static org.opensearch.common.util.FeatureFlags.REMOTE_PUBLICATION_EXPERIMENTAL;
+import static org.opensearch.gateway.remote.RemoteClusterStateService.REMOTE_CLUSTER_STATE_ENABLED_SETTING;
 import static org.opensearch.node.remotestore.RemoteStoreNodeAttribute.REMOTE_STORE_CLUSTER_STATE_REPOSITORY_NAME_ATTRIBUTE_KEY;
 import static org.opensearch.node.remotestore.RemoteStoreNodeAttribute.REMOTE_STORE_REPOSITORY_SETTINGS_ATTRIBUTE_KEY_PREFIX;
 import static org.opensearch.node.remotestore.RemoteStoreNodeAttribute.REMOTE_STORE_REPOSITORY_TYPE_ATTRIBUTE_KEY_FORMAT;
@@ -971,7 +973,7 @@ public class CoordinationStateTests extends OpenSearchTestCase {
             .put("node.attr." + REMOTE_STORE_CLUSTER_STATE_REPOSITORY_NAME_ATTRIBUTE_KEY, randomRepoName)
             .put(stateRepoTypeAttributeKey, FsRepository.TYPE)
             .put(stateRepoSettingsAttributeKeyPrefix + "location", "randomRepoPath")
-            .put(RemoteClusterStateService.REMOTE_CLUSTER_STATE_ENABLED_SETTING.getKey(), true)
+            .put(REMOTE_CLUSTER_STATE_ENABLED_SETTING.getKey(), true)
             .build();
 
         final CoordinationState coordinationState = createCoordinationState(persistedStateRegistry, node1, settings);
@@ -1000,6 +1002,16 @@ public class CoordinationStateTests extends OpenSearchTestCase {
             clusterStateCaptor.getValue().metadata().clusterUUIDCommitted(),
             equalTo(committedClusterState.metadata().clusterUUIDCommitted())
         );
+    }
+
+    public void testIsRemotePublicationEnabled_WithInconsistentSettings() {
+        // create settings with remote state disabled but publication enabled
+        Settings settings = Settings.builder()
+            .put(REMOTE_CLUSTER_STATE_ENABLED_SETTING.getKey(), false)
+            .put(REMOTE_PUBLICATION_EXPERIMENTAL, true)
+            .build();
+        CoordinationState coordinationState = createCoordinationState(psr1, node1, settings);
+        assertFalse(coordinationState.isRemotePublicationEnabled());
     }
 
     public static CoordinationState createCoordinationState(

